### PR TITLE
Bug fix/issue #9612

### PR DESCRIPTION
**Pull request by Dronplane**
Port of fix for issue #9612 from 3.4

Fixes ViewNode executor to not stop on first not-found-in-collection document. As index is eventually consistent - this is expected to happen sometimes. Verified by GTEST test.

Added index reversals for insert/update/remove for RocksDBCollection. Verified by js tests (only server side as test requires failpoints)

Changed index order storage from vector to set to allow indexes reordering. Now indexes that need reversal stored (and executed) after all other indexes to reduce possibility of reversals. Verified by GTEST test.

http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5796/

### CODE REVIEW

File: arangod/Aql/IResearchViewExecutor.cpp [hunk 1]
```diff
@@ -288,13 +288,14 @@ bool IResearchViewExecutorBase<Impl, Traits>::next(ReadContext& ctx) {
...
+     LOG_TOPIC("550cd", TRACE, arangodb::iresearch::TOPIC)
+            << "failed to write row in node executor";
   }
-
-  // no documents found, we're exhausted.
-  return false;
+  // return true anyway as we possbly failed to find one document 
+  // in collection due to eventual index consistency
+  // but there may be more other alive docs)
+  return true;
```
Comment (gnusi) [inline arangod/Aql/IResearchViewExecutor.cpp]: we'll illegally query affect stats

File: arangod/ClusterEngine/ClusterCollection.cpp [hunk 2]
```diff
@@ -384,9 +385,9 @@ bool ClusterCollection::dropIndex(TRI_idx_iid_t iid) {
 
   size_t i = 0;
```
Comment (gnusi) [inline arangod/ClusterEngine/ClusterCollection.cpp]: do we still need `i`?

File: arangod/StorageEngine/PhysicalCollection.h [hunk 3]
```diff
@@ -23,7 +23,7 @@
 
 #ifndef ARANGOD_VOCBASE_PHYSICAL_COLLECTION_H
 #define ARANGOD_VOCBASE_PHYSICAL_COLLECTION_H 1
-
+#include <set>
```
Comment (gnusi) [inline arangod/StorageEngine/PhysicalCollection.h]: I doubt we need this header as it has to be included in `PhysicalCollection.h`

**Comment by Dronplane** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5800/

File: arangod/ClusterEngine/ClusterCollection.cpp [hunk 4]
```diff
@@ -382,16 +383,14 @@ bool ClusterCollection::dropIndex(TRI_idx_iid_t iid) {
...
   }
 
-  size_t i = 0;
   WRITE_LOCKER(guard, _indexesLock);
-  for (std::shared_ptr<Index> index : _indexes) {
-    if (iid == index->id()) {
-      _indexes.erase(_indexes.begin() + i);
+  for (auto it  : _indexes) {
+    if (iid == it->id()) {
+      _indexes.erase(it);
```
Comment (jsteemann) [inline arangod/ClusterEngine/ClusterCollection.cpp]: Doesn't this invalidate the iterator `it` that is used in the `for` loop around?
Is that safe to do?

File: arangod/MMFiles/MMFilesCollection.cpp [hunk 5]
```diff
@@ -2462,27 +2460,19 @@ bool MMFilesCollection::dropIndex(TRI_idx_iid_t iid) {
...
-
-      _indexes.erase(_indexes.begin() + i);
-
       // update statistics
       MMFilesIndex* midx = static_cast<MMFilesIndex*>(idx.get());
       if (midx->isPersistent()) {
         --_persistentIndexes;
       }
-
+      _indexes.erase(idx);
```
Comment (jsteemann) [inline arangod/MMFiles/MMFilesCollection.cpp]: again, is this safe here? we are invalidating the iterator `idx`, which we use in the outer for loop.

File: arangod/MMFiles/MMFilesCollection.cpp [hunk 6]
```diff
@@ -2821,14 +2811,15 @@ Result MMFilesCollection::truncate(transaction::Methods& trx, OperationOptions&
...
-  auto indexes = _indexes;
-  size_t const n = indexes.size();
-
   TRI_voc_tick_t tick = TRI_NewTickServer();
-  for (size_t i = 1; i < n; ++i) {
-    auto idx = indexes[i];
-    TRI_ASSERT(idx->type() != Index::IndexType::TRI_IDX_TYPE_PRIMARY_INDEX);
-    idx->afterTruncate(tick);
+  if (_indexes.size() > 1) {
+    auto indexes = _indexes;
```
Comment (jsteemann) [inline arangod/MMFiles/MMFilesCollection.cpp]: why are we taking a copy of _indexes here? I think we can work directly on `_indexes`.

File: arangod/MMFiles/MMFilesCollection.cpp [hunk 7]
```diff
@@ -3262,34 +3253,32 @@ Result MMFilesCollection::insertSecondaryIndexes(arangodb::transaction::Methods&
   Result result;
 
   READ_LOCKER(guard, _indexesLock);
+  if (_indexes.size() > 1) {
+    auto indexes = _indexes;
```
Comment (jsteemann) [inline arangod/MMFiles/MMFilesCollection.cpp]: why are we copying `_indexes` here? can't we work on `_indexes` directly?

File: arangod/MMFiles/MMFilesCollection.cpp [hunk 8]
```diff
@@ -3311,26 +3300,23 @@ Result MMFilesCollection::deleteSecondaryIndexes(transaction::Methods& trx,
...
-      continue;
-    }
-
-    Result res = midx->remove(trx, documentId, doc, mode);
-
-    if (res.fail()) {
-      // an error occurred
-      result = res;
+  if (_indexes.size() > 1) {
+    auto indexes = _indexes;
```
Comment (jsteemann) [inline arangod/MMFiles/MMFilesCollection.cpp]: why are we copying `_indexes`? can't we work directly on `_indexes`?

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 9]
```diff
@@ -444,10 +446,10 @@ bool RocksDBCollection::dropIndex(TRI_idx_iid_t iid) {
   {
     size_t i = 0;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp]: I think variable `i` can now be removed here.

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 10]
```diff
@@ -1188,14 +1190,18 @@ void RocksDBCollection::figuresSpecific(std::shared_ptr<arangodb::velocypack::Bu
...
+  while (it != indexes.begin()) {
     it--;
     auto* rIdx = static_cast<RocksDBIndex*>(it->get());
-    if (rIdx->type() == Index::TRI_IDX_TYPE_IRESEARCH_LINK) {
-      std::forward<F>(op)(rIdx);
+    if (rIdx->needsReversal()) {
+      if (std::forward<F>(op)(rIdx).fail()) {
+        // best effort for reverse failed. Let`s trigger full rollback  
+        // or we will end up with inconsistent storage and indexes
+        THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp]: we should never throw TRI_ERROR_INTERNAL without a corresponding message attached.
I suggest changing this to `THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, "some descriptive message");`.

File: arangod/StorageEngine/PhysicalCollection.h [hunk 11]
```diff
@@ -101,9 +101,16 @@ class PhysicalCollection {
...
+  struct IndexOrder {
+    bool operator()(const std::shared_ptr<Index>& left,
+                              const std::shared_ptr<Index>& right) const;
+  };
+
+  using IndexContainerType = std::set<std::shared_ptr<Index>, IndexOrder> ;
   /// @brief find index by definition
   static std::shared_ptr<Index> findIndex(velocypack::Slice const&,
-                                          std::vector<std::shared_ptr<Index>> const&);
+                                         IndexContainerType const&);
```
Comment (jsteemann) [inline arangod/StorageEngine/PhysicalCollection.h]: indentation seems to be off by one byte.

File: arangod/ClusterEngine/ClusterCollection.cpp (hunk 4 above)
Comment (Dronplane) [reply arangod/ClusterEngine/ClusterCollection.cpp]: Yes, it invalidates. But it doesn`t matter as loop terminates after that.

File: arangod/MMFiles/MMFilesCollection.cpp (hunk 5 above)
Comment (Dronplane) [reply arangod/MMFiles/MMFilesCollection.cpp]: Looks safe to me as far as next line terminates the loop

File: arangod/ClusterEngine/ClusterCollection.cpp (hunk 4 above)
Comment (jsteemann) [reply arangod/ClusterEngine/ClusterCollection.cpp]: True

**Comment by Dronplane** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5803/

**Pull request merged**